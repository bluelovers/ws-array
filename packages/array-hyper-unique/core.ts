import { defaultFilter, equals, defaultChecker } from './lib/util';
import { IOptions } from './lib/types';

export * from './lib/types';
export * from './lib/util';

export function array_unique<T>(arr: T, options: IOptions<T> = {}): T
{
	if (!Array.isArray(arr))
	{
		throw new TypeError(`Expected an Array but got ${typeof arr}.`)
	}

	const cb = defaultFilter(options);

	if (options.overwrite)
	{
		let index = arr.length;

		while (index--)
		{
			let val = arr[index];

			if (!cb(val, index, arr))
			{
				arr.splice(index, 1);
			}
		}

		return arr;
	}

	// @ts-ignore
	return arr.filter(cb);
}

export function array_unique_overwrite<T>(arr: T, options: IOptions<T> = {}): T
{
	return array_unique(arr, {
		...options,
		overwrite: true,
	});
}

export function lazy_unique<T extends any[]>(arr: T): T
export function lazy_unique<T, T1, T2>(a1: T1, a2: T2, ...arr: T[]): Array<T | T1 | T2>
export function lazy_unique<T>(...arr: Array<T | T[]>): T | (T | T[])[]
// @ts-ignore
export function lazy_unique<T>(...arr: Array<T | T[]>)
{
	if (arr.length > 1)
	{
		return array_unique(arr);
	}

	return array_unique(arr[0]);
}

export function lazy_unique_overwrite<T>(...arr: Array<T | T[]>)
{
	if (arr.length > 1)
	{
		return array_unique_overwrite(arr);
	}

	return array_unique_overwrite(arr[0]);
}

lazy_unique.array_unique = array_unique;
lazy_unique.array_unique_overwrite = array_unique_overwrite;

lazy_unique.lazy_unique_overwrite = lazy_unique_overwrite;

lazy_unique.equals = equals;
lazy_unique.defaultFilter = defaultFilter;
lazy_unique.defaultChecker = defaultChecker;

lazy_unique.lazy_unique = lazy_unique;
lazy_unique.default = lazy_unique;

Object.defineProperty(lazy_unique, "__esModule", { value: true });

export default lazy_unique
