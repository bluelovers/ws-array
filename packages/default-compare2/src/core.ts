import { _typeOf } from './util';

export const enum EnumSortCompareResult
{
	KEEP = 0,
	RIGHT = 1,
	LEFT = -1,
}

export function numberCompare(a: unknown, b: unknown)
{
	if (a < b)
	{
		return EnumSortCompareResult.LEFT
	}
	else if (a > b)
	{
		return EnumSortCompareResult.RIGHT
	}

	return EnumSortCompareResult.KEEP
}

/**
 * will return undefined when a and b not null/undefined
 */
export function nullCompare(a: unknown, b: unknown)
{
	const typeA = _typeOf(a);
	const typeB = _typeOf(b);

	if (typeA === 'null')
	{
		return typeB === 'null' ? EnumSortCompareResult.KEEP : (typeB === 'undefined'
			? EnumSortCompareResult.LEFT
			: EnumSortCompareResult.RIGHT);
	}
	else if (typeA === 'undefined')
	{
		return typeB === 'null' ? EnumSortCompareResult.RIGHT : (typeB === 'undefined'
			? EnumSortCompareResult.KEEP
			: EnumSortCompareResult.RIGHT);
	}
	else if (typeB === 'null' || typeB === 'undefined')
	{
		return EnumSortCompareResult.LEFT;
	}
}

export function symbolCompare(a: unknown, b: unknown)
{
	const typeA = _typeOf(a);
	const typeB = _typeOf(b);

	if (typeA === 'symbol')
	{
		return typeB === 'symbol' ? numberCompare(String(a), String(b)) : EnumSortCompareResult.LEFT
	}
	else if (typeB === 'symbol')
	{
		return EnumSortCompareResult.RIGHT
	}

	return numberCompare(a, b)
}

export function dateCompare(a: Date, b: Date)
{
	return a.getTime() - b.getTime()
}

export function stringCompareLocale(a: string, b: string)
{
	return a.localeCompare(b)
}
