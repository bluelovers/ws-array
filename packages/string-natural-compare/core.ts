/**
 * Created by user on 2020/6/4.
 */

import _naturalCompare from 'string-natural-compare';

export interface IOptionsStringNaturalCompare
{
	/**
	 * Set to true to compare strings case-insensitively. Default: false.
	 */
	caseInsensitive?: boolean,
	/**
	 * A string of characters that define a custom character ordering. Default: undefined.
	 */
	alphabet?: string,
}

export interface IOptionsNaturalCompare extends IOptionsStringNaturalCompare
{
	desc?: boolean,
}

/**
 * Compare alphanumeric strings the same way a human would,
 * using a natural order algorithm
 * (originally known as the alphanum algorithm)
 * where numeric characters are sorted
 * based on their numeric values rather than their ASCII values.
 */
export function naturalCompare(a, b, opts?: IOptionsNaturalCompare)
{
	let i: number;

	if (typeof a === 'number' && typeof b === 'number')
	{
		i = a - b
	}
	else
	{
		if (typeof a === 'number')
		{
			a = String(a);
		}
		else if (typeof b === 'number')
		{
			b = String(b);
		}

		if (a === b)
		{
			return 0
		}

		i = _naturalCompare(a, b, opts)
	}

	if (i !== 0 && opts?.desc)
	{
		i = 0 - i;
	}

	return i
}

/**
 * create compare with preset options
 */
export function createNew(opts?: IOptionsNaturalCompare)
{
	return (a, b) => naturalCompare(a, b, opts)
}

/**
 * compare strings case-insensitively
 */
export const compareCaseInsensitive = createNew({
	caseInsensitive: true,
})

naturalCompare.createNew = createNew;
naturalCompare.compareCaseInsensitive = compareCaseInsensitive;
naturalCompare.default = naturalCompare;

export default naturalCompare
