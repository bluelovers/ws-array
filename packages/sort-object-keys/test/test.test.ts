/**
 * Created by user on 2018/2/4/004.
 */

import { sortObject } from '../core';
import sortObject2 from '../';
import sortObject3 from '../index';
import sortObject4 from '../core';

sortObject([]);
sortObject2([]);
sortObject3([]);
sortObject4([]);

it(`sort object`, function ()
{
	let actual = sortObject({
		c: 1,
		b: 1,
		d: 1,
		a: 1,
	});

	let expected = {
		a: 1,
		b: 1,
		c: 1,
		d: 1,
	};

	expect(actual).toStrictEqual(expected);
	strictStringify(actual, expected);

	strictKeys(actual, expected);

	expect(Object.keys(actual)).toMatchSnapshot();
	expect(actual).toMatchSnapshot();
});

it('sort object v2', function ()
{
	let actual = sortObject({
		c: 1,
		b: 1,
		d: 1,
		a: 1,
	}, {
		keys: ['b', 'c', 'a', 'd'],
		onlyKeys: true,
	})

	expect(Object.keys(actual)).toStrictEqual(['b', 'c', 'a', 'd']);
	expect(Object.keys(actual)).toMatchSnapshot();
	expect(actual).toMatchSnapshot();
});

it(`source will not change order`, function ()
{
	let source = { a: 1, c: 2, e: 5, d: 4, b: 3 };
	let actual = sortObject(source, {
		keys: ['a', 'b'],
	});
	let expected = { a: 1, b: 3, c: 2, d: 4, e: 5 };

	expect(actual).toStrictEqual(expected);
	expect(actual).toStrictEqual(source);

	strictKeys(actual, source, true);
	strictKeys(actual, expected);

	expect(actual === source).toStrictEqual(false);

	expect(Object.keys(actual)).toMatchSnapshot();
	expect(actual).toMatchSnapshot();
});

it(`source will change order`, function ()
{
	let source = { a: 1, c: 2, e: 5, d: 4, b: 3 };
	let actual = sortObject(source, {
		keys: ['a', 'b'],
		useSource: true,
	});
	let expected = { a: 1, b: 3, c: 2, d: 4, e: 5 };

	expect(actual).toStrictEqual(expected);
	expect(actual).toStrictEqual(source);

	strictKeys(actual, source);
	strictKeys(actual, expected);

	expect(actual === source).toStrictEqual(true);

	expect(Object.keys(actual)).toMatchSnapshot();
	expect(actual).toMatchSnapshot();
});

it('should not add extra keys to object', function ()
{
	let actual = sortObject({
		b: 1,
		a: 1,
	}, ['a', 'b', 'c', 'd'])

	expect(Object.keys(actual)).toMatchSnapshot();
	expect(Object.keys(actual)).toStrictEqual([
		'a',
		'b',
	]);
	expect(actual).toMatchSnapshot();
});

it('should follow sort function', function ()
{
	let actual = sortObject({
		"key-1": 1,
		"key-3": 1,
		"key-10": 1,
		"key-2": 1,
	}, removeKeyAncCompareIndex)

	strictKeys(actual, {
		"key-1": 1,
		"key-2": 1,
		"key-3": 1,
		"key-10": 1,
	});

	expect(Object.keys(actual)).toMatchSnapshot();
	expect(actual).toMatchSnapshot();

	function removeKeyAncCompareIndex(keyA, keyB)
	{
		let a = parseInt(keyA.slice(4));
		let b = parseInt(keyB.slice(4));
		return a - b;
	}
});

it('should add extra keys to object', function ()
{
	let actual = sortObject({
		b: 1,
		a: 1,
	}, {
		keys: ['a', 'b', 'c', 'd'],
		allowNotExists: true,
	})

	expect(Object.keys(actual)).toMatchSnapshot();

	expect(Object.keys(actual)).toStrictEqual(['a', 'b', 'c', 'd']);
	expect(actual).toMatchSnapshot();
});

it('with onlyKeys', function ()
{
	let source = {
		c: 1,
		b: 1,
		d: 1,
		a: 1,
	};
	let actual = sortObject(source, {
		keys: ['b', 'd'],
		onlyKeys: true,
		useSource: true,
	})

	expect(Object.keys(actual)).toMatchSnapshot();
	expect(Object.keys(source)).toMatchSnapshot();

	expect(Object.keys(actual)).toStrictEqual(['b', 'd']);
	expect(Object.keys(source)).toStrictEqual(['c', 'b', 'd', 'a']);
	expect(actual).toMatchSnapshot();
});

it('with onlyKeys v2', function ()
{
	let source = {
		c: 1,
		b: 1,
		d: 1,
		a: 1,
	};

	expect(() => sortObject(source, {
		onlyKeys: true,
	})).toThrow()
});

it('with desc', function ()
{
	let actual = sortObject({
		c: 1,
		b: 1,
		d: 1,
		a: 1,
	}, {
		keys: ['b', 'c', 'a', 'd'],
		desc: true,
	})

	let keys = Object.keys(actual);
	expect(keys).toMatchSnapshot();

	expect(keys).toStrictEqual(['d', 'a', 'c', 'b']);
	expect(actual).toMatchSnapshot();
});

it('sort object v3', function ()
{
	let actual = sortObject({
		c: 1,
		b: 1,
		d: 1,
		a: 1,
	}, ['b', 'c', 'a', 'd'])

	expect(Object.keys(actual)).toStrictEqual(['b', 'c', 'a', 'd']);
	expect(Object.keys(actual)).toMatchSnapshot();
	expect(actual).toMatchSnapshot();
});

function strictStringify(actual, expected, not?: boolean)
{
	let ex = expect(JSON.stringify(actual))
	return (not ? ex.not : ex).toStrictEqual(JSON.stringify(expected));
}

function strictKeys(actual, expected, not?: boolean)
{
	let ex = expect(Object.keys(actual));
	return (not ? ex.not : ex).toStrictEqual(Object.keys(expected));
}
