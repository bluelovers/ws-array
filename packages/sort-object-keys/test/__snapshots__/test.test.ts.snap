// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should add extra keys to object 1`] = `
Array [
  "a",
  "b",
  "c",
  "d",
]
`;

exports[`should add extra keys to object 2`] = `
Object {
  "a": 1,
  "b": 1,
  "c": undefined,
  "d": undefined,
}
`;

exports[`should follow sort function 1`] = `
Array [
  "key-1",
  "key-2",
  "key-3",
  "key-10",
]
`;

exports[`should follow sort function 2`] = `
Object {
  "key-1": 1,
  "key-10": 1,
  "key-2": 1,
  "key-3": 1,
}
`;

exports[`should not add extra keys to object 1`] = `
Array [
  "a",
  "b",
]
`;

exports[`should not add extra keys to object 2`] = `
Object {
  "a": 1,
  "b": 1,
}
`;

exports[`sort object 1`] = `
Array [
  "a",
  "b",
  "c",
  "d",
]
`;

exports[`sort object 2`] = `
Object {
  "a": 1,
  "b": 1,
  "c": 1,
  "d": 1,
}
`;

exports[`sort object v2 1`] = `
Array [
  "b",
  "c",
  "a",
  "d",
]
`;

exports[`sort object v2 2`] = `
Object {
  "a": 1,
  "b": 1,
  "c": 1,
  "d": 1,
}
`;

exports[`sort object v3 1`] = `
Array [
  "b",
  "c",
  "a",
  "d",
]
`;

exports[`sort object v3 2`] = `
Object {
  "a": 1,
  "b": 1,
  "c": 1,
  "d": 1,
}
`;

exports[`source will change order 1`] = `
Array [
  "a",
  "b",
  "c",
  "d",
  "e",
]
`;

exports[`source will change order 2`] = `
Object {
  "a": 1,
  "b": 3,
  "c": 2,
  "d": 4,
  "e": 5,
}
`;

exports[`source will not change order 1`] = `
Array [
  "a",
  "b",
  "c",
  "d",
  "e",
]
`;

exports[`source will not change order 2`] = `
Object {
  "a": 1,
  "b": 3,
  "c": 2,
  "d": 4,
  "e": 5,
}
`;

exports[`with desc 1`] = `
Array [
  "d",
  "a",
  "c",
  "b",
]
`;

exports[`with desc 2`] = `
Object {
  "a": 1,
  "b": 1,
  "c": 1,
  "d": 1,
}
`;

exports[`with onlyKeys 1`] = `
Array [
  "b",
  "d",
]
`;

exports[`with onlyKeys 2`] = `
Array [
  "c",
  "b",
  "d",
  "a",
]
`;

exports[`with onlyKeys 3`] = `
Object {
  "b": 1,
  "d": 1,
}
`;
